jdb/assignment3/wordsearch.py

```diff
@@ -0,0 +1,138 @@
+        path[(x, y)] = True
+        backtrack_request = yield path
+        if not backtrack_request:
+          yield from self._traverse(path)
+        path.popitem()
+    else:
+      for dx, dy in product([-1, 0, 1], [-1, 0, 1]):
+        (xl, yl), _ = path.popitem()
+        x, y = xl + dx, xl + dy
+        path[(xl, yl)] = True
+        if (0 <= x < self.x_max and 0 <= y < self.y_max
+            and (x, y) not in path):
+          path[(x, y)]= True
+          backtrack_request = yield path
+          if not backtrack_request:
+            yield from self._traverse(path)
+          path.popitem()
+
+
+def load_lexicon():
+  trie = Trie()
+  # On linux, install the file below with:
+  # $ sudo apt-get install wamerican
+  with open('/usr/share/dict/american-english') as f:
+    for word in [w.strip().lower() for w in f
+                 if not w.strip().endswith("'s") and len(w.strip())>1]:
+      trie.add(word)
+  return trie
+
+
+class Trie:
+
+  def __init__(self):
+    self.children = {}
+    self.full_word = False
+
+  def add(self, letters):
+    first, *rest = letters
+    if first not in self.children:
+      self.children[first] = Trie()
+    if rest:
+      self.children[first].add(rest)
+    else:
+      self.children[first].full_word = True
+
+  def __getitem__(self, letters):
+    first, *rest = letters
+    if first not in self.children:
+      return False
+    return self.children[first][rest] if rest else self.children[first]
+
+  def is_word(self, letters):
+    trie = self[letters]
+    return trie.full_word if trie else False
+
+  def is_prefix(self, letters):
+    return bool(self[letters])
+
+
+# run the unittest with python -m unittest wordsearch.py
+import unittest
+
+class TestTrie(unittest.TestCase):
+
+  def test_trie(self):
+    trie = Trie()
+    for word in 'car', 'card', 'cart', 'cat':
+      trie.add(word)
+    for word in 'car', 'card', 'cart', 'cat':
+      self.assertTrue(trie.is_word(word))
+      self.assertTrue(trie.is_prefix(word))
+    for word in 'c', 'ca':
+      self.assertFalse(trie.is_word(word))
+      self.assertTrue(trie.is_prefix(word))
+    self.assertFalse(trie['NoTfOuNd'])
+
+
+class TestGrid(unittest.TestCase):
+
+  def test_traverse(self):
+    neighbors = set(
+      tuple(p) for p in Grid('abc', 'def', 'ghi', lexicon=True)._traverse())
+    self.assertIn(((1, 1),), neighbors)
+    self.assertIn(((0, 0),), neighbors)
+    self.assertIn(((1, 2),), neighbors)
+    self.assertNotIn(((1, -1),), neighbors)
+
+  def test_getitem(self):
+    grid = Grid('abc', 'def', 'ghi', lexicon=True)
+    for letter, position in zip('abcdefghi',
+                                ((x, y) for y in range(3) for x in range(3))):
+      self.assertEqual(grid[position], letter)
+
+  def test_words(self):
+    words = Grid('aar', 'tcd').words()
+    for word in ['card', 'data', 'act', 'arc', 'cad', 'car', 'cat', 'rat',
+                 'rca', 'tad', 'tar', 'ac', 'ad']:
+      self.assertIn(word, words)
```

Comment:
:slightly_frowning_face: Unittest give the following error:

**FAIL: test_words (wordsearch.TestGrid)
Traceback (most recent call last):
  File "/mnt/FileStorage/Projects_progs/ForVacancies/CodeU-2018-Group2/jdb/assignment3/wordsearch.py", line 138, in test_words
    self.assertIn(word, words)
AssertionError: 'data' not found in {'tad', 'cat', 'tc', 'tar', 'cr', 'act', 'rd', 'rca', 'card', 'ta', 'ar', 'car', 'arc', 'cd', 'cad', 'ad', 'ca', 'at', 'ac'}**

Then if I'm deleting the 'data', it shows the same error for 'rat'. The rest of words are found correctly.
What's wrong? :thinking: 
:slightly_frowning_face: Unittest give the following error:

**FAIL: test_words (wordsearch.TestGrid)
Traceback (most recent call last):
  File "/mnt/FileStorage/Projects_progs/ForVacancies/CodeU-2018-Group2/jdb/assignment3/wordsearch.py", line 138, in test_words
    self.assertIn(word, words)
AssertionError: 'data' not found in {'tad', 'cat', 'tc', 'tar', 'cr', 'act', 'rd', 'rca', 'card', 'ta', 'ar', 'car', 'arc', 'cd', 'cad', 'ad', 'ca', 'at', 'ac'}**

Then if I'm deleting the 'data', it shows the same error for 'rat'. The rest of words are found correctly.
What's wrong? :thinking: Both words 'data' and 'rat' contains diagonal step, but 'rca', for example, contains it too and not causing the error.